# cosmo port of the sqlite shell: /zip/.args support for white
# labeling, symbolic %t strace logging, errno hygiene after sqliterc,
# and a no-op shell_stifle_history since cosmo linenoise has no
# linenoiseHistorySetMaxLen.
--- a/third_party/sqlite3/shell.c
+++ b/third_party/sqlite3/shell.c
@@ -156,6 +156,9 @@
 #include <math.h>
 #include <stdint.h>
 #include "third_party/sqlite3/sqlite3.h"
+#include "libc/cosmo.h"                     /* [jart] LoadZipArgs() */
+#include "libc/runtime/runtime.h"           /* [jart] __strace */
+#include "libc/runtime/symbols.internal.h"  /* [jart] GetSymbolTable() */
 typedef sqlite3_int64 i64;
 typedef sqlite3_uint64 u64;
 typedef unsigned char u8;
@@ -211,7 +214,7 @@
 # define shell_add_history(X) linenoiseHistoryAdd(X)
 # define shell_read_history(X) linenoiseHistoryLoad(X)
 # define shell_write_history(X) linenoiseHistorySave(X)
-# define shell_stifle_history(X) linenoiseHistorySetMaxLen(X)
+# define shell_stifle_history(X) /* [jart] not in cosmo linenoise */
 # define shell_readline(X) linenoise(X)
 
 #else
@@ -26464,6 +26467,7 @@
 
       if( bVerbose ){
         const char *zCand = sqlite3_expert_report(p,0,EXPERT_REPORT_CANDIDATES);
+        if( zCand==0 ) zCand = ""; /* [cosmo] gcc -Wformat-overflow */
         cli_puts("-- Candidates -----------------------------\n", out);
         cli_printf(out, "%s\n", zCand);
       }
@@ -36261,6 +36265,7 @@
   if( sqliterc != sqliterc_override ){
     sqlite3_free(sqliterc);
   }
+  errno = 0; /* [jart] clear error */
 }
 
 /*
@@ -36510,6 +36515,15 @@
   char **azCmd = 0;
   int *aiCmd = 0;
   const char *zVfs = 0;           /* Value of -vfs command-line option */
+
+  /* [jart] ensure %t symbols in strace log are symbolic */
+  if( __strace > 0 ){
+    GetSymbolTable();
+  }
+
+  /* [jart] support /zip/.args file for white labeling */
+  LoadZipArgs(&argc, &argv);
+
   setvbuf(stderr, 0, _IONBF, 0); /* Make sure stderr is unbuffered */
 
 #ifdef SQLITE_SHELL_FIDDLE
